proto/substrait/extensions/extensions.proto

```diff
@@ -67,6 +76,23 @@ message SimpleExtensionDeclaration {
     // more than one impl per name in the YAML.
     string name = 3;
   }
+
+  message TypeParameter {
```

Comment:
I think this should be part of the type variable as opposed to being normalized in the extensiontypevariation. 

It means we should probably deprecate this line:
https://github.com/substrait-io/substrait/blob/main/proto/substrait/type.proto#L39

``` 
uint32 user_defined_type_reference = 31;
```

And instead introduce something more like 
```
UserDefinedType user_defined = 32;

message UserDefinedType {
  uint32 user_defined_type_reference = 1;

  // Parameterization, if this is a compound type variation.
  repeated TypeParameter parameters = 2; 

  message TypeParameter {
    oneof parameter {
      // Explicitly null/unspecified parameter, to select the default value (if
      // any).
      google.protobuf.Empty null = 1;

      // Data type parameters, like the i32 in LIST<i32>.
      Type data_type = 2;

      // Value parameters, like the 10 in VARCHAR<10>.
      bool boolean = 3;
      int64 integer = 4;
      string enum = 5;
      string string = 6;
    }
  }
}
```

This would make user defined definitions consistent with all other compound types, defining them within the operations rather than having to constantly normalize them. E.g. `creating list<i8>` would be basically the same as creating `mylist<i8>`.

Reply:
That's a good point. I did it this way for consistency with function overloads (i.e. you'd need separate anchors for each overload), to make it a bit easier to compare types, and to save some repetition in the plan, but yours is probably the stronger argument.

Reply:
Any status update on the comments here?

Reply:
I was just spitballing with this and haven't looked at this for a while to be honest, but I can pick it back up next week.